Remove incorrect base_link rotation.

gavanderhoorn opened the pull request:
This fixes the long-standing problem with `base_link` being incorrectly rotated 180 degree over Z wrt the real robot (ie: "the connector was sticking out the wrong side").

This commit removes that rotation, while also making sure `base` keeps its correct location and orientation.

The `base->tool0` transform should be unaffected, as that had a 180 degree counter-rotation.

gavanderhoorn commented:
@fmauch: could I ask you to verify this against an e-Series robot?

@fmessmer: this should resolve your #503.

gavanderhoorn commented:
A good test would be to run `rosrun tf2_tools echo.py base tool0` and comparing that with what the TP shows (for an all-zeros toolframe of course).

fmauch commented:
> @fmauch: could I ask you to verify this against an e-Series robot?

sure thing, I'll try to squeeze it in tomorrow.

fmauch commented:
I can confirm, that on our UR16e the TP and tf_echo `base->tool0` and `base->tool0_controller` all report the same values. :-)

gavanderhoorn commented:
@fmauch wrote:
> I can confirm, that on our UR16e the TP and tf_echo `base->tool0` and `base->tool0_controller` all report the same values. :-)

Nice, thanks.

Seeing as the xacro macro is identical across all supported robot models, that should fix it for non e-Series as well.

gavanderhoorn commented:
While it's nice to have removed the rotation, the issue seems to be a bit deeper than I initially thought.

The current xacro macro has the robot stretched out over X- when in its zero-pose. While by convention (and REP-103), it should be pointing in the X+ direction.

In essence, the robot is now modelled *backwards*.

I'll see how I can best rectify this, but in any case it won't affect the `base->tool0` transform, as that is already correct and a few extra rotations of `base_link` will not change that (provided we counter-rotate `base`).

fmessmer commented:
well, while this PR aligns the cable bit of the `base` mesh with the y-axis of the `base` link - as it should be, it also changes the orientation of `base_link` which might cause trouble for people using the UR as part of a more complex scene or robot

yes, it is stated in the `macro` with a comment
and yes, there are breaking changes announced for melodic

but still, what I tried to bring up in #503 was way less intrusive, by simply adjusting the `origin` tag of `${prefix}base_link` (or rotate the meshes) - this would not have changed the tf tree

gavanderhoorn commented:
> well, while this PR aligns the cable bit of the `base` mesh with the y-axis of the `base` link - as it should be, it also changes the orientation of `base_link` which might cause trouble for people using the UR as part of a more complex scene or robot
> 
> yes, it is stated in the `macro` with a comment
> and yes, there are breaking changes announced for melodic

Exactly, so this is the point in time to do this.

I'm sorry for users who will have to update their models, but this is a `-devel` branch and nothing has ever been released for melodic.

> but still, what I tried to bring up in #503 was way less intrusive, by simply adjusting the `origin` tag of `${prefix}base_link` (or rotate the meshes) - this would not have changed the tf tree

The current model is wrong. It needs to be fixed.

gavanderhoorn committed:
description: remove the 'base_correction' property.

There is an offset between 'base_link' and 'base', but it's a fixed one, known and does not need to be a property.

Rectify the problem with the 'base_link' mesh in a follow-up commit.

gavanderhoorn committed:
description: introduce base_link_inertia link.

To work around the root-link-with-inertia limitation in KDL.

gavanderhoorn committed:
description: rotate for UR internal frames.

gavanderhoorn committed:
description: fix orientation of 'base' frame.

So it's coincident with the robot's/controller's 'Base' frame. It's actually 180 degrees rotated over Z wrt 'base_link'.

gavanderhoorn committed:
description: correct orientation of Base mesh.

Connector should be aligned with Y+ of 'base' frame.

gavanderhoorn commented:
Ok @fmessmer: I've used a different approach.

`base_link` remains as it is now (it was already REP-103 aligned), and I've introduced a new frame (`base_link_inertia`) which both serves to rotate things around to have X+ pointing "backward" (for the rest of the chain) as well as to deal with the `root-link-with-inertia` limitation in KDL.

Orientation of the `base` mesh is fixed with this, and the model is now a bit more coherent (although there are still rotations around Z of `pi` radians in several places).

Together with #506 I believe this now gives us a proper model, with the expected frames.

As always: `base->tool0` should return whatever the TP shows for that transform.

Let me know whether this works for you.

@fmauch: if you could check this *one last time* ?

fmauch commented:
And again, it reports the same values :-)

gavanderhoorn commented:
> And again, it reports the same values :-)

Good to hear :)

gavanderhoorn commented:
Travis is doing weird things again. Two PR builds apparently for one PR from one branch. And both have already completed, but the second one has not reported to GH it has completed ..

gavanderhoorn commented:
@fmessmer: I'm going to merge this by eod.

If you have any further comments not addressed then, we can fix those in subsequent PRs.

gavanderhoorn commented:
Just going to wait for Travis to be happy and then I'll merge this.

gavanderhoorn commented:
Thanks @fmauch for the early review and thanks @fmessmer for the additional input.

fmessmer commented:
sorry, for not getting back earlier....but the merged version works for me (only tested briefly)

zeal-up commented:
@gavanderhoorn 
Sorry, I still a little confusing about your patch.
In my understanding, the problems we should solves are: 1. make the cable point along the positive y axis. 2. the transfrom from tool0->base is the true tcp pose of the real robot(the pose read frm TP).

The introduction of the base_link_inertia frame is necessary, but even with your patch, the robot in rviz still has a rotation issue, it has a 180 degree rotation compare with the real robot.

So, why don't you align base_link and base frame? a.k,a, the joint between base and base_link is a fix joint with no rotation.
The fix procedures is as below(after your PR):
1. align base and base_link frame
```
<joint name="${prefix}base_link-base_fixed_joint" type="fixed">
      <!-- Note the rotation over Z of pi radians: as base_link is REP-103
           aligned (ie: has X+ forward, Y+ left and Z+ up), this is needed
           to correctly align 'base' with the 'Base' coordinate system of
           the UR controller.
      -->
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <parent link="${prefix}base_link"/>
      <child link="${prefix}base"/>
    </joint>
```

2. set the rotation between base_link and base_link_inertia to be zero:
```
<joint name="${prefix}base_link-base_link_inertia" type="fixed">
      <parent link="${prefix}base_link" />
      <child link="${prefix}base_link_inertia" />
      <!-- 'base_link' is REP-103 aligned (so X+ forward), while the internal
           frames of the robot/controller have X+ pointing backwards.
           Use the joint between 'base_link' and 'base_link_inertia' (a dummy
           link/frame) to introduce the necessary rotation over Z (of pi rad).
      -->
      <origin xyz="0 0 0" rpy="0 0 0" />
    </joint>
```

All have be down.

Maybe I miss something, but it's pretty good in all my test:
1. the cable point along the positive y-axis
2. the tool0->base transform match the TP in the robot
3. the visualization is exactly match the real robot
